basic_module/distribution/src/lib.rs

```diff
@@ -0,0 +1,294 @@
+                current_reward: self.current_reward.clone(),
+                calculated_reward: self.calculated_reward.clone(),
+                total_min_fee: self.total_min_fee,
+                min_fee: self.min_fee,
+            }
+            .into(),
+        );
+    }
+
+    pub fn update_quantity(&mut self, address: &Public, quantity: u64) {
+        if quantity == 0 {
+            return
+        }
+        *self.current_reward.entry(*address).or_insert(0) = quantity;
+    }
+
+    pub fn drain_calculated_rewards(&mut self) -> StateResult<BTreeMap<Public, u64>> {
+        let mut drained = BTreeMap::new();
+        mem::swap(&mut drained, &mut self.calculated_reward);
+
+        self.update_intermediate_rewards();
+        Ok(drained)
+    }
+
+    pub fn drain_current_rewards(&mut self) -> StateResult<BTreeMap<Public, u64>> {
+        let mut drained = BTreeMap::new();
+        mem::swap(&mut drained, &mut self.current_reward);
+
+        self.update_intermediate_rewards();
+        Ok(drained)
+    }
+}
+
+impl Default for IntermediateRewards {
+    fn default() -> Self {
+        IntermediateRewards {
+            current_reward: BTreeMap::default(),
+            calculated_reward: BTreeMap::default(),
+            total_min_fee: Default::default(),
+            min_fee: Default::default(),
+        }
+    }
+}
+
+impl From<Vec<u8>> for IntermediateRewards {
+    fn from(_vec: Vec<u8>) -> Self {
+        IntermediateRewards::default()
+    }
+}
+
+impl From<IntermediateRewards> for Vec<u8> {
+    fn from(_reward: IntermediateRewards) -> Self {
+        Vec::default()
+    }
+}
+
+pub struct FeeDistributeIter<'a> {
+    total_stakes: u64,
+    total_min_fee: u64,
+    remaining_fee: u64,
+    stake_holdings: hash_map::Iter<'a, Public, u64>,
+}
+
+impl<'a> FeeDistributeIter<'a> {
+    pub fn remaining_fee(&self) -> u64 {
+        self.remaining_fee
+    }
+}
+
+pub fn fee_distribute<S: ::std::hash::BuildHasher>(
+    total_min_fee: u64,
+    stakes: &HashMap<Public, u64, S>,
+) -> FeeDistributeIter<'_> {
+    FeeDistributeIter {
+        total_stakes: stakes.values().sum(),
+        total_min_fee,
+        remaining_fee: total_min_fee,
+        stake_holdings: stakes.iter(),
+    }
+}
+
+impl<'a> Iterator for FeeDistributeIter<'a> {
+    type Item = (&'a Public, u64);
+    fn next(&mut self) -> Option<(&'a Public, u64)> {
+        if let Some((stakeholder, stake)) = self.stake_holdings.next() {
+            let share = share(self.total_stakes, *stake, self.total_min_fee);
+            self.remaining_fee = self.remaining_fee.checked_sub(share).expect("Remaining fee shouldn't be depleted");
+            Some((stakeholder, share))
+        } else {
+            None
+        }
+    }
+}
+
+fn share(total_stakes: u64, stake: u64, total_min_fee: u64) -> u64 {
+    assert!(total_stakes >= stake);
+    u64::try_from((u128::from(total_min_fee) * u128::from(stake)) / u128::from(total_stakes)).unwrap()
+}
+
+pub struct ActionDataKeyBuilder {
```

Comment:
We don't need `ActionDataKeyBuilder` because when you request `set` to `substorageAccess` the key will be prefixed in the host side so that entries will not be overlapped with other modules.

Reply:
I removed `ActionDataKeyBuilder` and modified the login in `get_key()` to return `H256::default()`.

Reply:
Instead of the `H256::default()` you can define any key value like "intermediate_reward" inside the module. The only required trait bound is `AsRef<[u8]>`.